src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,630 @@
+            (void) consumed;  // prevent unused variable warning
+            assert(consumed == 1);
+            qd_log(tls->log_source,
+                   QD_LOG_TRACE,
+                   "[C%" PRIu64 "] qd_tls_encrypt pn_tls_give_encrypt_input_buffers success",
+                   tls->conn_id);
+
+            //
+            // Process TLS.
+            //
+            if (!process_tls(tls)) {
+                take_back_input_encrypt_buffs(tls);
+                return QD_TLS_ERROR;
+            }
+        }
+
+        // We are immediately taking back the  encrypt input buffers which means the tls encrypt input buffer capacity
+        // is restored.
+        take_back_input_encrypt_buffs(tls);
+        tls->tls_has_output = pn_tls_is_encrypt_output_pending(tls->tls_session);
+    }
+give_encrypt_output_buffers:
+    while (pn_tls_need_encrypt_output_buffers(tls->tls_session)) {
+        //
+        // We will give just one result buffer
+        //
+        // This is the raw buffer that will hold the encrypted results
+        pn_raw_buffer_t      encrypted_result_raw_buffer;
+        qd_adaptor_buffer_t *adaptor_buff = qd_adaptor_buffer_raw(&encrypted_result_raw_buffer);
+        (void) adaptor_buff;  // prevent unused variable warning
+
+        //
+        // Send the empty raw_buffers to proton tls. Proton tls will put the encrypted data into
+        // encrypted_result_raw_buffers
+        //
+        size_t encrypt_result_buffers_count =
+            pn_tls_give_encrypt_output_buffers(tls->tls_session, &encrypted_result_raw_buffer, 1);
+        assert(encrypt_result_buffers_count == 1);
+        (void) encrypt_result_buffers_count;  // prevent unused variable warning
+
+        qd_log(tls->log_source,
+               QD_LOG_TRACE,
+               "[C%" PRIu64 "] qd_tls_encrypt pn_tls_give_encrypt_output_buffers success",
+               tls->conn_id);
+
+        //
+        // Process TLS.
+        //
+        if (!process_tls(tls)) {
+            take_back_input_encrypt_buffs(tls);
+            return QD_TLS_ERROR;
+        }
+
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] qd_tls_encrypt process_tls successful", tls->conn_id);
+    }
+
+    pn_raw_buffer_t encrypted_raw_output_buff;
+    while (pn_tls_take_encrypt_output_buffers(tls->tls_session, &encrypted_raw_output_buff, 1)) {
+        qd_adaptor_buffer_t *encrypted_adaptor_buff =
+            qd_get_adaptor_buffer_from_pn_raw_buffer(&encrypted_raw_output_buff);
+        DEQ_INSERT_TAIL(*encrypted_buffs, encrypted_adaptor_buff);
+        qd_log(tls->log_source,
+               QD_LOG_TRACE,
+               "[C%" PRIu64 "] qd_tls_encrypt pn_tls_take_encrypt_output_buffers, encrypted_raw_output_buff.size=%zu",
+               tls->conn_id,
+               encrypted_raw_output_buff.size);
+    }
+
+    if (pn_tls_need_encrypt_output_buffers(tls->tls_session)) {
+        goto give_encrypt_output_buffers;
```

Comment:
This goto needs to go, too.

Reply:
After our discussion, you mentioned it is ok to keep the goto statements, so I have not removed them.